### DISCUSSION TITLE
Resetting values in repeated elements

### DISCUSSION THREAD

**Discussion by farmaazon** [slint-ui/slint#3400] [Q&A] [answered]
My problem is very similar to [this discussion](https://github.com/slint-ui/slint/discussions/3108) but with addition of repeated elements.

I've created [an example](https://slintpad.com/?snippet=import+%7B+Button%2C+HorizontalBox%2C+Slider%2C+VerticalBox+%7D+from+%22std-widgets.slint%22%3B%0A%0A%0Aexport+component+Demo+%7B%0A++++property+%3C%5B%7Bx%3A+float%2C+y%3Afloat%7D%5D%3E+prop%3A+%5B%0A++++++++%7B+x%3A+1.0%2C+y%3A+2.0+%7D%2C+%0A++++++++%7B+x%3A+2.0%2C+y%3A+3.0+%7D%2C%0A++++++++%7B+x%3A+3.0%2C+y%3A+4.0+%7D%2C%0A++++++++%7B+x%3A+4.0%2C+y%3A+1.0+%7D%2C%0A++++++++%5D%3B%0A%0A++++HorizontalBox+%7B%0A++++++++VerticalBox+%7B%0A++++++++++++Slider+%7B%0A+++++++++++++++width%3A+100px%3B%0A++++++++++++++++minimum%3A+0.0%3B%0A++++++++++++++++maximum%3A+4.0%3B%0A++++++++++++++++value%3A+prop%5B0%5D.x%3B%0A%0A++++++++++++++++changed%28value%29+%3D%3E+%7B%0A++++++++++++++++++++prop%5B0%5D.x+%3D+value%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A%0A++++++++++++Button+%7B%0A++++++++++++++++text%3A+%22Reset%22%3B%0A%0A++++++++++++++++clicked+%3D%3E+%7B%0A++++++++++++++++++++prop+%3D+%5B%7B+x%3A+0.0%2C+y%3A+0.0%7D%2C+%7B+x%3A+0.0%2C+y%3A+0.0+%7D%5D%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++%7D%0A%0A++++++++VerticalBox+%7B%0A++++++++++++for+elem%5Bindex%5D+in+prop%3A+HorizontalLayout+%7B%0A++++++++++++++++Slider+%7B%0A++++++++++++++++++++width%3A+100px%3B%0A++++++++++++++++++++minimum%3A+0.0%3B%0A++++++++++++++++++++maximum%3A+4.0%3B%0A++++++++++++++++++++value%3A+elem.x%3B%0A%0A++++++++++++++++++++changed%28value%29+%3D%3E+%7B%0A++++++++++++++++++++++++prop%5Bindex%5D.x+%3D+value%3B%0A++++++++++++++++++++%7D%0A++++++++++++++++%7D%0A%0A++++++++++++++++Slider+%7B%0A++++++++++++++++++++width%3A+100px%3B%0A++++++++++++++++++++minimum%3A+0.0%3B%0A++++++++++++++++++++maximum%3A+4.0%3B%0A++++++++++++++++++++value%3A+elem.y%3B%0A%0A++++++++++++++++++++changed%28value%29+%3D%3E+%7B%0A++++++++++++++++++++++++prop%5Bindex%5D.y+%3D+value%3B%0A++++++++++++++++++++%7D%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++%7D%0A++++%7D%0A%7D%0A) where I have a bunch of sliders bound to some array model. The reset button changes the model. As  one could  expect, changing any slider destroys binding, so pressing reset after moving slider does not update it.

Now, in case of single slider, I can live with setting its value manually after pressing Reset. I don't know however, what can I do with repeated elements:
1. I cannot iterate over them setting their values. Or can I?
2. I cannot add binding inside repeated element to an external callback (button.clicked in this case). Or can I?

Is there anything I can do except replacing "for" with bunch of "ifs" or waiting for two-way binding to model support?

**Comment by ogoffart** [comment]
I think in this case, this also exhibit bug https://github.com/slint-ui/slint/issues/2598

**Reply by farmaazon** [reply]
I realized they're re-used, but I thought it's rather a feature than a bug :smile:

**Comment by farmaazon** [comment, marked as answer]
I develop a dirty workaround, posting it here in case anybody would be interested:

1. In some accessible place (e.g. global singleton) declare a bool property (let's say "resetting") and a callback ("reset")
2. Implement the callback in Rust: set the property to true and start single-shot timer with 0 duration which will set it back to false.
3. make the element containing repeated elements conditional: `if !resetting: `
4. Call the callback in places where you want to restore bindings.

The UI will blink for a moment, but bindings will be restored. And I guess it is far from best performance, therefore I can't wait for two-way bindings!
